### PR TITLE
KIALI-2269 Change model to dissociate runtimes vs dashboards

### PR DESCRIPTION
**Pull request by jotak**
JIRA https://issues.jboss.org/browse/KIALI-2269

One runtime (ex: Vert.x) may have several dashboards (ex: Vert.x Server, Vert.x Event bus, etc.)
This is useful in order not to load too many metrics at the same time, which can be very penalizing for kiali response time

Side changes:
- Split our temporary vert.x dashboard into smaller ones - adding many charts
- Don't use maps to extract unique elements from aggregated lists, but keep using lists instead to preserve ordering. Note that perf impact will be negligible because we don't expect more than a few elements in lists.
- Rename annotation kiali.io/dashboards -> kiali.io/runtimes

### CODE REVIEW

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#427](https://paste.fedoraproject.org/paste/Pzj2~pqoOet4nCv2FKtcgQ)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1030]](https://paste.fedoraproject.org/paste/Q5iQh9EQBikGtTjR5UcqMg)

**Comment by lucasponce** [comment]
Travis catches that swagger.json needs to be updated.

**Comment by lucasponce** [comment]
@jotak do you remember the steps to setup locally a scenario ?
I remember there was a QE JIRA, but same steps apply here as well ?

**Comment by jotak** [comment]
@lucasponce swagger updated

In order to test, you can run this:

```
oc new-project mesh-arena
wget https://raw.githubusercontent.com/jotak/demo-mesh-arena/master/full-metrics.yml
oc apply -f <(istioctl kube-inject -f full-metrics.yml)
oc expose service ui
```

It won't work ootb with istio 1.0.x (should work on 1.1 branch though I didn't tested) so you need to update prometheus' configmap and restart it. Here's the configmap I use: https://paste.fedoraproject.org/paste/vTwYAIulrGcvFKrcO9Ew-w
Just the two last jobs, `kubernetes-pods` and `kubernetes-pods-istio-secure`, are modified.

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#433](https://paste.fedoraproject.org/paste/JqbBnXHmd2GAiPkgT0nj3Q)
  * :red_circle: run-kiali-e2e-tests [#[1038]](https://paste.fedoraproject.org/paste/V9zXP77woXKKhDQdtNrb9Q)

File: business/apps.go [hunk 1]
```diff
@@ -200,22 +200,17 @@ func fetchNamespaceApps(k8s kubernetes.IstioClientInterface, namespace string, a
...
-			for _, ref := range pod.CustomDashboards {
-				if ref != "" {
-					uniqueRefs[ref] = ref
-				}
-			}
-		}
+		allPods = append(allPods, workload.Pods...)
 	}
+	uniqueRefsList := getUniqueRuntimes(allPods)
 	mon, err := kubernetes.NewKialiMonitoringClient()
```
Comment (lucasponce) [inline business/apps.go]: If we create a client on every call, perhaps we are hitting some performance issues as well.
I think that in the business layer we can re-use some of existing clients.

File: business/dashboards.go [hunk 2]
```diff
@@ -161,25 +161,60 @@ func (in *DashboardsService) GetIstioDashboard(params prometheus.IstioMetricsQue
...
+	wg.Wait()
+
+	runtimes := []models.Runtime{}
+	for _, dashboard := range dashboards {
+		if dashboard == nil {
+			continue
+		}
+		runtime := getDashboardRuntime(dashboard)
+		ref := models.DashboardRef{
+			Template: dashboard.Metadata["name"].(string),
```
Comment (lucasponce) [inline business/dashboards.go]: Can this cast fail ?

File: kubernetes/kiali_monitoring_types.go [hunk 3]
```diff
@@ -19,7 +19,8 @@ var kialiMonitoringGroupVersion = schema.GroupVersion{
 }
 
 type MonitoringDashboard struct {
-	Spec MonitoringDashboardSpec
+	Metadata map[string]interface{}
```
Comment (lucasponce) [inline kubernetes/kiali_monitoring_types.go]: If you need it, you can map the Metadata with the Kubernetes Metadata type to get all the mappings done.
I think CRDs follow same type (if I'm now wrong here).
Probably you can save a couple of checks, but it's also ok to leave it as it is here.

File: business/apps.go (hunk 1 above)
Comment (jotak) [reply business/apps.go]: iirc in the business layer we also create new clients on every calls, or did it change?

File: business/dashboards.go (hunk 2 above)
Comment (jotak) [reply business/dashboards.go]: I'm not sure kube API would accept something other than a string for name in a first place, but sure, I can check for conversion

**Comment by lucasponce** [comment]
@hhovsepy I have some issues with my local machine, would you mind to run a quick smoke test on this PR and its UI ? Thanks

**Comment by jotak** [comment]
@hhovsepy I've added some debug logs, could you retry with this new version and show the server logs when a workload details is displayed? Thanks

**Comment by hhovsepy** [comment]
``
I0124 19:47:48.302781       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.302809       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.302837       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.302863       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.344716       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.387739       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.387822       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.387851       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.387868       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.387884       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.403652       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.505316       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.533953       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.560893       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.602791       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:47:48.603184       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:47:48.615889       1 pod.go:100] Pod runtimes annotation not found
I0124 19:47:48.652025       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:47:48.652165       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:47:48.698760       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool mesh-arena]
I0124 19:47:48.761020       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool]
I0124 19:47:48.782077       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:23.592645       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.592771       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.592916       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.593144       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.593308       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool mesh-arena]
I0124 19:48:23.593438       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool]
I0124 19:48:23.593567       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:23.866769       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.866905       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.872322       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:23.872462       1 pod.go:98] Found pod runtimes annotation: [vertx-client vertx-pool]
I0124 19:48:24.305739       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool]
I0124 19:48:24.313033       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-client vertx-pool mesh-arena]
I0124 19:48:24.394147       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:25.848546       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:25.870271       1 pod.go:98] Found pod runtimes annotation: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:25.885132       1 workloads.go:998] Unique runtimes: [vertx-server vertx-eventbus vertx-pool]
I0124 19:48:25.886352       1 dashboards.go:171] Loading dashboard: vertx-pool
I0124 19:48:25.886437       1 dashboards.go:171] Loading dashboard: vertx-server
I0124 19:48:25.886679       1 dashboards.go:171] Loading dashboard: vertx-eventbus
E0124 19:48:25.940973       1 dashboards.go:174] Cannot get dashboard vertx-pool in namespace mesh-arena. Error was: the server could not find the requested resource (get monitoringdashboards.monitoring.kiali.io)
E0124 19:48:25.941226       1 dashboards.go:174] Cannot get dashboard vertx-server in namespace mesh-arena. Error was: the server could not find the requested resource (get monitoringdashboards.monitoring.kiali.io)
E0124 19:48:25.942174       1 dashboards.go:174] Cannot get dashboard vertx-eventbus in namespace mesh-arena. Error was: the server could not find the requested resource (get monitoringdashboards.monitoring.kiali.io)
I0124 19:48:25.942288       1 dashboards.go:211] Returning runtimes: []
``

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#446](https://paste.fedoraproject.org/paste/RaGJghknKlLFmSORxMkgcQ)
  * :red_circle: run-kiali-e2e-tests [#[1058]](https://paste.fedoraproject.org/paste/bEaCIArO5Nx5nBr7gvCOTQ)

**Comment by jotak** [comment]
@hhovsepy I've been trying again this morning and found something weird in QE's environment:

1. the mesh arena's pods are deployed both in 'mesh-arena' namespace but also in 'bookinfo' namespace
2. when coming from the overview and clicking on mesh-arena card, the namespace filtering doesn't work: we still see items from bookinfo and others
3. so I believe what we saw yesterday (or what we didn't see: no dashboard) was because we were looking at the wrong namespace. I tried today, and I can see the dashboards.

So IMO there's no bug on this PR, but there's an unrelated bug about namespace selection (we should check on master).

**Comment by hhovsepy** [comment]
Thank you @jotak 
I can see the metrics now.
screenshot from 2019-01-25 13-08-30
screenshot from 2019-01-25 13-09-00
screenshot from 2019-01-25 13-09-28

**Comment by jotak** [comment]
Thanks. I removed the debug-logs commit

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#459](https://paste.fedoraproject.org/paste/3cCcgSN9DtMIHw~hzjNVwg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1073]](https://paste.fedoraproject.org/paste/zwFWdzuTnApimerRM3PmDw)

**Pull request merged**